File: internal/authz/permissions.go
package authz

import "Sixth_world_Suday/internal/role"

type Permission string

const (
	PermAll               Permission = "*"
	PermViewAdminPanel    Permission = "view_admin_panel"
	PermViewStats         Permission = "view_stats"
	PermViewAuditLog      Permission = "view_audit_log"
	PermManageSettings    Permission = "manage_settings"
	PermManageRoles       Permission = "manage_roles"
	PermDeleteAnyUser     Permission = "delete_any_user"
	PermBanUser           Permission = "ban_user"
	PermViewUsers         Permission = "view_users"
	PermDeleteAnyPost     Permission = "delete_any_post"
	PermDeleteAnyComment  Permission = "delete_any_comment"
	PermEditAnyPost       Permission = "edit_any_post"
	PermEditAnyComment    Permission = "edit_any_comment"
	PermManageVanityRoles Permission = "manage_vanity_roles"
	PermManageBannedWords Permission = "manage_banned_words"
	PermResetPassword     Permission = "reset_password"
	PermManageChannels    Permission = "manage_channels"
)

var rolePermissions = map[role.Role][]Permission{
	RoleSuperAdmin: {
		PermAll,
	},
	RoleAdmin: {
		PermAll,
	},
	RoleModerator: {
		PermViewAdminPanel,
		PermViewStats,
		PermViewUsers,
		PermDeleteAnyPost,
		PermDeleteAnyComment,
		PermEditAnyPost,
		PermEditAnyComment,
		PermBanUser,
		PermManageChannels,
	},
}

File: frontend/src/utils/permissions.ts
export type SiteRole = "super_admin" | "admin" | "moderator" | "gm";

export const ROLE_GROUPS: { role: SiteRole; label: string }[] = [
    { role: "super_admin", label: "Owners" },
    { role: "admin", label: "Admins" },
    { role: "moderator", label: "Moderators" },
    { role: "gm", label: "Hosts" },
];

export function isSiteStaff(role: SiteRole | undefined | null): boolean {
    return role === "super_admin" || role === "admin" || role === "moderator";
}

export type Permission =
    | "ban_user"
    | "manage_roles"
    | "view_admin_panel"
    | "manage_settings"
    | "view_audit_log"
    | "view_stats"
    | "view_users"
    | "delete_any_user"
    | "delete_any_post"
    | "delete_any_comment"
    | "edit_any_post"
    | "edit_any_comment"
    | "manage_vanity_roles"
    | "manage_banned_words"
    | "manage_channels"
    | "reset_password"
    | "lock_files";

const rolePermissions: Record<string, Permission[]> = {
    super_admin: [
        "ban_user",
        "manage_roles",
        "view_admin_panel",
        "manage_settings",
        "view_audit_log",
        "view_stats",
        "view_users",
        "delete_any_user",
        "delete_any_post",
        "delete_any_comment",
        "edit_any_post",
        "edit_any_comment",
        "manage_vanity_roles",
        "manage_banned_words",
        "manage_channels",
        "reset_password",
        "lock_files",
    ],
    admin: [
        "ban_user",
        "manage_roles",
        "view_admin_panel",
        "manage_settings",
        "view_audit_log",
        "view_stats",
        "view_users",
        "delete_any_user",
        "delete_any_post",
        "delete_any_comment",
        "edit_any_post",
        "edit_any_comment",
        "manage_vanity_roles",
        "manage_banned_words",
        "manage_channels",
        "reset_password",
        "lock_files",
    ],
    moderator: [
        "delete_any_post",
        "delete_any_comment",
        "edit_any_post",
        "edit_any_comment",
        "view_admin_panel",
        "view_stats",
        "view_users",
        "ban_user",
        "manage_channels",
        "lock_files",
    ],
    gm: ["lock_files"],
};

export function can(role: SiteRole | undefined, perm: Permission): boolean {
    if (!role) {
        return false;
    }
    return rolePermissions[role]?.includes(perm) ?? false;
}

export function canAccessAdmin(role: SiteRole | undefined): boolean {
    return can(role, "view_admin_panel");
}
